src/vs/workbench/contrib/debug/browser/debugSession.ts

```diff
@@ -302,7 +307,7 @@ export class DebugSession implements IDebugSession {
 
 			this.initialized = true;
 			this._onDidChangeState.fire();
-			this.debugService.setExceptionBreakpoints((this.raw && this.raw.capabilities.exceptionBreakpointFilters) || []);
+			this.exceptionBreakpoints = this.debugService.createExceptionBreakpoints((this.raw && this.raw.capabilities.exceptionBreakpointFilters) || []);
```

Comment:
I don't think it's necessary to call all the way into the DebugModel here to create real ExceptionBreakpoint objects. I think it's ok for the DebugSession to expose raw exceptionBreakpointFilters and we can create real objects when that session is selected.

Reply:
Seems that many things exposed by the DebugSession are raw DebugProtocol types

Reply:
The alternative would be to keep a map from session to exception breakpoints in model or somewhere else, if we ultimately want different breakpoint configuration per session, no?

Theoretically, in this case, session owns the breakpoints, so made sense to keep it in session.

Model has the info required for creating the real ExceptionBreakpoint objects, i.e. values from last/current session to be used as default.

Agreed that the current flow is not intuitive. Suggestions?

Reply:
I see that the enablement state of the exception breakpoint is on that object, so those need to live in some map somewhere. I think that putting them on the session is not right, because if I enable "caught exceptions" for one node debug session, that should be enabled for all node debug sessions. So you could have a map from debug adapter type to exception filters for that type.

But if I am debugging an electron app, you have node and chrome configs, those DAs contribute the same breakpoint filters, and if I enable "caught exceptions" for node, I want it to also be enabled for chrome. We could keep a set of all exception breakpoint filters each with the debug types that they are enabled for. I think we should actually be working towards something like that.

`ExceptionBreakpointsFilter` has an ID. If an unrelated DA contributes an exception filter with the same ID but a different label, should it be treated the same as the node/chrome filter? Or should we look at all fields of `ExceptionBreakpointsFilter` to decide whether it's a match?

Reply:
> I see that the enablement state of the exception breakpoint is on that object, so those need to live in some map somewhere. I think that putting them on the session is not right, because if I enable "caught exceptions" for one node debug session, that should be enabled for all node debug sessions. So you could have a map from debug adapter type to exception filters for that type.

Having exception breakpoints in session, including enablement also fixes the scenario mentioned in #147096. But to be fair, if we want enablement different for every session, it could apply to other breakpoints as well, depending on the particular multi-session scenario. Though with this restoring earlier preferences doesn't work so well. Primarily because we'll anyway have to choose which session to store if multiple session from the same DA are active. To store different preferences for different DAs would also be challenging this case. 

> But if I am debugging an electron app, you have node and chrome configs, those DAs contribute the same breakpoint filters, and if I enable "caught exceptions" for node, I want it to also be enabled for chrome. We could keep a set of all exception breakpoint filters each with the debug types that they are enabled for. I think we should actually be working towards something like that.

> ExceptionBreakpointsFilter has an ID. If an unrelated DA contributes an exception filter with the same ID but a different label, should it be treated the same as the node/chrome filter? Or should we look at all fields of ExceptionBreakpointsFilter to decide whether it's a match?

What if we keep the exception breakpoints separate for each DA always, even if the id and name are same? We could show the breakpoints from all sessions, along with the DA name it is coming from to differentiate those. Which would also reduce confusion since the breakpoint pane won't change as per the active session. The user gets the choice to select exception breakpoints per DA. Since the two sessions are using different DAs, it is also likely that the significance and usage of the exception breakpoints can also be different, even if the id/name are same.

The original changes I made was somewhat similar to this, where the user sees all the exception breakpoints at once (though it did not differentiate per DA as such, it only looked at the id). The original changes also gave better experience when it comes to saving and restoring the breakpoints across multiple DAs, with a twist that if some EB from an inactive DA that are no longer needs to be saved, they needed to be manually deleted by the user. I feel something like that would give a much better experience when working with multiple debuggers at once. I can try this approach again if it makes sense.

Reply:
> We could show the breakpoints from all sessions, along with the DA name it is coming from to differentiate those.

Strong disagree on that, that will be a lot of noise. It should act as a global view that updates for the active session, same as the debug toolbar.

> Since the two sessions are using different DAs, it is also likely that the significance and usage of the exception breakpoints can also be different, even if the id/name are same.

If this is about the node/chrome case, I disagree on that too. It's also sort of an issue that it's not necessarily obvious which sessions are "node" and which are "chrome" when debugging electron.

I think that the direction we are going in now is definitely the right one. We just need to decide whether exception breakpoint state is shared across DA types for the same ID (or ID + name, or something). I would like to say yes. It would be interesting to see whether any of the popular debug extensions have collisions in their exception filter IDs, I have no clue.

Reply:
If we want to share EB state across sessions, session can no longer own it, which is the case with the current changes.

Here is what I'm thinking, though, cannot figure out a couple of things.

Model can keep a list of all exception breakpoints across currently active DAs. When a new session starts, we add exception breakpoints for it to the model if not already exists with same id/name. Model can then expose a method to get exception breakpoints for a particular session for the UI to display.

When do we remove the old exception breakpoints from the model in this case?

We could remove breakpoints from the model when the session ends, if no other active session supports it. If we do this, we regress the current behaviour of saving the enablement for exception breakpoints across restarts of a session.

We "could" never remove it. In that case the enablement is remembered even for multi-debugger cases as well, which is great. But the model would bloat over time in theory.

Btw, finding a design that can save enablement even for multi-debugger scenarios would help a lot. It is annoying that the exception breakpoints would fallback to default every time I start debugging with a multi-debugger setup.

Suggestions?

Reply:
What you are describing is pretty much what I have in mind. I think that just looking at the ID is probably ok (worth taking a quick look at popular DA extensions) and saving the selection in workspace state forever is also fine.

Reply:
Just looking at the id won't be sufficient since the breakpoints can have different support for conditions along with different names and descriptions.

For example, Java and Nodejs debugger has an overlapping id, but with different condition support and different descriptions.

I've updated the PR with changes, have a look?

Reply:
> For example, Java and Nodejs debugger has an overlapping id, but with different condition support and different descriptions.

Interesting, what is that ID?

Reply:
nodejs provides 'uncaught' and 'all', Java provides 'uncaught' and 'caught'. 'uncaught' is the conflicting one for these two.

Since the name are generic, I'd guess more debuggers would have similar conflicting names.